src/olympia/lib/crypto/signing.py

```diff
@@ -0,0 +1,278 @@
+import hashlib
+import os
+import shutil
+import tempfile
+import zipfile
+
+from base64 import b64decode, b64encode
+
+from django.conf import settings
+from django.core.files.storage import default_storage as storage
+from django.utils.encoding import force_bytes
+
+import requests
+import six
+import waffle
+
+from django_statsd.clients import statsd
+from requests_hawk import HawkAuth
+from signing_clients.apps import JarExtractor
+from asn1crypto import cms
+
+import olympia.core.logger
+
+from olympia import amo
+
+
+log = olympia.core.logger.getLogger('z.crypto')
+
+
+SIGN_FOR_APPS = (amo.FIREFOX.id, amo.ANDROID.id)
+
+
+class SigningError(Exception):
+    pass
+
+
+def supports_firefox(file_obj):
+    """Return True if the file supports Firefox or Firefox for Android.
+
+    We only sign files that are at least compatible with Firefox/Firefox for
+    Android.
+    """
+    apps = file_obj.version.apps.all()
+    return apps.filter(max__application__in=SIGN_FOR_APPS)
+
+
+def get_id(addon):
+    """Return the addon GUID if <= 64 chars, or its sha256 hash otherwise.
+
+    We don't want GUIDs longer than 64 chars: bug 1203365.
+    """
+    guid = addon.guid
+    if len(guid) <= 64:
+        return guid
+    return hashlib.sha256(guid).hexdigest()
+
+
+def call_signing(file_obj):
+    """Get the jar signature and send it to the signing server to be signed."""
+    # Extract jar signature.
+    conf = settings.AUTOGRAPH_CONFIG
+    log.debug('Calling autograph service: {0}'.format(
+        conf['server_url']))
+
+    if waffle.sample_is_active('activate-autograph-file-signing'):
+        with storage.open(file_obj.current_file_path) as fobj:
+            input_data = b64encode(fobj.read())
+
+        signing_request = [{
+            'input': input_data,
+            'keyid': conf['signer'],
+            'options': {
+                'id': get_id(file_obj.version.addon),
+                # "Add-on variant A params (PKCS7 SHA1 and COSE ES256) work in
+                # Fx <57, so we can switch to that without breaking backwards
+                # compatibility"
+                # https://github.com/mozilla/addons-server/issues/9308
+                # This means, the pkcs7 sha1 signature is used for backwards
+                # compatibility and cose sha256 will be used for newer
+                # Firefox versions.
+                'pkcs7_digest': 'SHA1',
+                'cose_algorithms': ['ES256']
+            },
+        }]
+
+        with statsd.timer('services.sign.addon.autograph'):
+            response = requests.post(
+                '{server}/sign/file'.format(server=conf['server_url']),
+                json=signing_request,
+                auth=HawkAuth(id=conf['user_id'], key=conf['key']))
```

Comment:
cc @g-k to just double-check the code above and that the docs are right.

Reply:
:+1: lgtm

Also, not sure how much detail you want in include about signature validation in the docs, but:

* the relevant pref is "security.signed_app_signatures.policy"
* in Fx nightly (and stable I believe) it's currently set to 2 for COSEAndPKCS7WithSHA1OrSHA256 in https://searchfox.org/mozilla-central/rev/dac799c9f4e9f5f05c1071cba94f2522aa31f7eb/security/manager/ssl/tests/unit/test_signed_apps.js#190-198